stablehlo/reference/Ops.cpp

```diff
@@ -429,50 +429,63 @@ Tensor evalSliceOp(const Tensor &operand, Index startIndices, Sizes strides,
 SmallVector<Tensor> evalSortOp(ArrayRef<Tensor> inputs, Axis dimension,
                                bool isStable, Region &comparator,
                                Scope &scope) {
-  SmallVector<Tensor> runtimeResults;
-  for (auto input : inputs) runtimeResults.push_back(Tensor(input.getType()));
-  if (dimension < 0) dimension += inputs[0].getRank();
-  Index indices(inputs.front().getShape()[dimension]);
-  Sizes inputShape(inputs.front().getShape());
-  inputShape[dimension] = 1;
-  for (auto offsetIt = IndexSpaceIterator(inputShape, Index(inputShape.size()));
-       offsetIt != IndexSpaceIterator(inputShape, std::nullopt); ++offsetIt) {
-    std::iota(indices.begin(), indices.end(), 0);
-    auto compare = [&](int64_t lhs, int64_t rhs) {
+  SmallVector<Tensor> results;
+  for (auto input : inputs) results.push_back(Tensor(input.getType()));
+  auto adjustedDimension =
+      dimension >= 0 ? dimension : dimension + inputs[0].getRank();
+
+  for (auto resultIt = results[0].index_begin();
+       resultIt != results[0].index_end(); ++resultIt) {
+    // resultIt iterates through all indices in the index space, but sorting
+    // only needs to be done once per slice.
+    if ((*resultIt)[adjustedDimension] != 0) continue;
```

Comment:
I'd propose that we use an existing iterator instead of making up a new one. It's not perfect for the task (we need this weird continue), but making up a synthetic shape with a dummy dimension size is not perfect either.